src/compas/datastructures/assembly/assembly.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+
+from ..datastructure import Datastructure
+from ..network import Network
+
+
+class Assembly(Datastructure):
+    """A data structure for managing the connections between different parts of an assembly.
+
+    Attributes
+    ----------
+    attributes: dict
+        General attributes of the assembly that will be included in the data dict.
+    network: :class:`compas.datastructures.Network`
+        The network that is used under the hood to store the parts and their connections.
+    """
+
+    @property
+    def DATASCHEMA(self):
+        import schema
+        return schema.Schema({
+            "network": Network,
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        return 'assembly'
+
+    def __init__(self, *args, **kwargs):
+        super(Assembly, None).__init__(*args, **kwargs)
+        self.attributes = {}
+        self.network = Network()
+
+    def __str__(self):
+        tpl = "<Assembly with {} parts and {} connections>"
+        return tpl.format(self.network.number_of_nodes(), self.network.number_of_edges())
+
+    @property
+    def data(self):
+        """dict : A data dict representing the assembly data structure for serialization.
+        """
+        data = {
+            'attributes': self.attributes,
+            'network': self.network,
+        }
+        return data
+
+    @data.setter
+    def data(self, data):
+        self.attributes.update(data['attributes'] or {})
+        self.network = data['network']
+
+    def parts(self):
+        """The parts of the assembly."""
+        return self.network.nodes()
+
+    def connections(self):
+        """The connections between the parts."""
+        return self.network.edges()
+
+    def add_part(self, part, id=None, **kwargs):
+        """Add a part to the assembly."""
+        self.network.add_node(key=id, part=part, **kwargs)
```

Comment:
I don't follow... is the `part` a network node (as is returned by `self.parts`)? or something else?  And if `id` isn't specified, shouldn't this return the identifier that is generated?

Reply:
yep, `Part` is the new name for what we used to call `Element` in other assembly structures, so they are stored as network nodes.

And you're right about the need to return the generated identifier.

Reply:
Ah, maybe I misunderstood your question. Are you asking why `.parts()` will return identifiers instead of the `part` attribute stored in the node?

Reply:
it all doesn't really make sense to me yet either :) will have to try this on a few concrete examples first to get it right...

Reply:
Exactly.

Reply:
can a part be anything, or is it also an assembly (or a "part" which is a subclass of "assembly", since perhaps there should be only 1 "assembly" in an assembly) with at least one node?

Reply:
I would generally argue that the `Assembly` stuff should lean towards being normative/opinionated, unlike `Network` or `Mesh` that are more general.
So, a `Part` should not be anything, there should be a `Part` base class.
And I have the feeling that it is `Assembly` that needs to sub-class `Part` and not the other way around.

Also, we def need a way to lookup parts quickly, so, a `.part(id)` method or even an dict-like accessor to get to parts quick would be important.